src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -520,7 +519,7 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
         }
         def checkOverrideTerm() {
           other.cookJavaRawInfo() // #2454
-          if (!overridesTypeInPrefix(lowType, highType, rootType)) { // 8
+          if (!overridesTypeInPrefix(lowType, highType, rootType, low.isModuleOrModuleClass || high.isModuleOrModuleClass)) { // 8
```

Comment:
D'oh! This needs to be `low && high` of course, not `low || high`. Those extra tests are already paying off...